src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -64,30 +61,48 @@ public CursorsService(final ZooKeeperHolder zkHolder,
                           final EventTypeRepository eventTypeRepository,
                           final ZooKeeperLockFactory zkLockFactory,
                           final ZkSubscriptionClientFactory zkSubscriptionClientFactory,
-                          final SubscriptionKafkaClientFactory subscriptionKafkaClientFactory,
                           final CursorTokenService cursorTokenService) {
         this.zkHolder = zkHolder;
         this.topicRepository = topicRepository;
         this.subscriptionRepository = subscriptionRepository;
         this.eventTypeRepository = eventTypeRepository;
         this.zkLockFactory = zkLockFactory;
         this.zkSubscriptionClientFactory = zkSubscriptionClientFactory;
-        this.subscriptionKafkaClientFactory = subscriptionKafkaClientFactory;
         this.cursorTokenService = cursorTokenService;
     }
 
-    public Map<SubscriptionCursor, Boolean> commitCursors(final String subscriptionId,
+    public Map<SubscriptionCursor, Boolean> commitCursors(final String streamId, final String subscriptionId,
                                                           final List<SubscriptionCursor> cursors)
             throws NakadiException, InvalidCursorException {
 
         final Subscription subscription = subscriptionRepository.getSubscription(subscriptionId);
-        createSubscriptionInZkIfNeeded(subscription);
+        final ZkSubscriptionClient subscriptionClient = zkSubscriptionClientFactory.createZkSubscriptionClient(
+                subscription.getId());
+
+        validateCursors(subscriptionClient, cursors, streamId);
 
         return cursors.stream().collect(Collectors.toMap(Function.identity(),
                 Try.<SubscriptionCursor, Boolean>wrap(cursor -> processCursor(subscriptionId, cursor))
                         .andThen(Try::getOrThrow)));
     }
 
+    private void validateCursors(final ZkSubscriptionClient subscriptionClient,
+                                 final List<SubscriptionCursor> cursors,
+                                 final String streamId) throws InvalidCursorException
+    {
+        final Map<Partition.PartitionKey, Partition> partitions = Arrays.stream(subscriptionClient.listPartitions())
+                .collect(Collectors.toMap(Partition::getKey, Function.identity()));
+        final List<SubscriptionCursor> invalidCursors = cursors.stream().filter(cursor -> Try.cons(() ->
+                eventTypeRepository.findByName(cursor.getEventType())).getO().map(eventType -> {
+            final Partition.PartitionKey key = new Partition.PartitionKey(eventType.getTopic(), cursor.getPartition());
+            final Partition partition = partitions.get(key);
+            return partition == null || !streamId.equals(partition.getSession());
```

Comment:
And last thing - calls to listSessions() and listPartitions() should be made under `zkClient.runLocked()`, because they must operate with consistent data.